ISPN-5451 data container striping

wburns opened the pull request:
https://issues.jboss.org/browse/ISPN-5451
https://issues.jboss.org/browse/ISPN-9212
https://issues.jboss.org/browse/ISPN-9245
https://issues.jboss.org/browse/ISPN-9246

Initial on heap implementation

tristantarrant commented:
I'd like to see both performance and allocation details of this change.

wburns commented:
Looks like it breaks a bunch of hibernate 2LC tests

https://ci.infinispan.org/job/Infinispan/job/PR-5997/3/#showFailuresLink

wburns commented:
run performance tests please

wburns commented:
run performance tests please library

wburns commented:
Very quickly changing the https://github.com/infinispan/infinispan-benchmarks/tree/master/iteration to do a test for inserts with already precomputed keys and values I got the following:

```
SEGMENTED

Benchmark          (keyObjectSize)  (nodes)  (useIdentityCache)  (useStrings)  (valueObjectSize)   Mode  Cnt     Score    Error  Units
JMHBenchmarks.put               10        6                true         false               1000  thrpt   20  8839.924 ± 69.350  ops/s

Benchmark                                           (keyObjectSize)  (nodes)  (useIdentityCache)  (useStrings)  (valueObjectSize)   Mode  Cnt      Score       Error   Units
JMHBenchmarks.put                                                10        6                true         false               1000  thrpt   20   8587.319 ±    70.790   ops/s
JMHBenchmarks.put:·gc.alloc.rate                                 10        6                true         false               1000  thrpt   20    130.952 ±    22.624  MB/sec
JMHBenchmarks.put:·gc.alloc.rate.norm                            10        6                true         false               1000  thrpt   20  24037.155 ±  3990.484    B/op
JMHBenchmarks.put:·gc.churn.PS_Eden_Space                        10        6                true         false               1000  thrpt   20    133.564 ±   283.264  MB/sec
JMHBenchmarks.put:·gc.churn.PS_Eden_Space.norm                   10        6                true         false               1000  thrpt   20  24622.623 ± 52220.116    B/op
JMHBenchmarks.put:·gc.churn.PS_Survivor_Space                    10        6                true         false               1000  thrpt   20      0.031 ±     0.085  MB/sec
JMHBenchmarks.put:·gc.churn.PS_Survivor_Space.norm               10        6                true         false               1000  thrpt   20      5.809 ±    15.653    B/op
JMHBenchmarks.put:·gc.count                                      10        6                true         false               1000  thrpt   20      3.000              counts
JMHBenchmarks.put:·gc.time                                       10        6                true         false               1000  thrpt   20     16.000                  ms


MASTER

Benchmark          (keyObjectSize)  (nodes)  (useIdentityCache)  (useStrings)  (valueObjectSize)   Mode  Cnt     Score     Error  Units
JMHBenchmarks.put               10        6                true         false               1000  thrpt   20  8688.200 ± 171.789  ops/s


Benchmark                                           (keyObjectSize)  (nodes)  (useIdentityCache)  (useStrings)  (valueObjectSize)   Mode  Cnt      Score       Error   Units
JMHBenchmarks.put                                                10        6                true         false               1000  thrpt   20   8763.348 ±    53.413   ops/s
JMHBenchmarks.put:·gc.alloc.rate                                 10        6                true         false               1000  thrpt   20    133.793 ±    22.774  MB/sec
JMHBenchmarks.put:·gc.alloc.rate.norm                            10        6                true         false               1000  thrpt   20  24022.719 ±  3991.557    B/op
JMHBenchmarks.put:·gc.churn.PS_Eden_Space                        10        6                true         false               1000  thrpt   20    133.562 ±   283.259  MB/sec
JMHBenchmarks.put:·gc.churn.PS_Eden_Space.norm                   10        6                true         false               1000  thrpt   20  24132.248 ± 51179.819    B/op
JMHBenchmarks.put:·gc.churn.PS_Survivor_Space                    10        6                true         false               1000  thrpt   20      0.014 ±     0.053  MB/sec
JMHBenchmarks.put:·gc.churn.PS_Survivor_Space.norm               10        6                true         false               1000  thrpt   20      2.445 ±     9.496    B/op
JMHBenchmarks.put:·gc.count                                      10        6                true         false               1000  thrpt   20      3.000              counts
JMHBenchmarks.put:·gc.time                                       10        6                true         false               1000  thrpt   20     16.000                  ms

```

This was found with 2 runs of jmh for master and this branch. One without and one with `-prof gc` set. The results seem pretty consistent within each other.

From the code the only difference for a `Cache.put` and `Cache.get` should be one additional single segment lookup and array access for a given key.

wburns commented:
I also did a 1 minute sampling of constant puts in the same environment with flight recorder and saw results that were quite similar

10 GCs for both averaging at 6.5ms each in duration.

The total TLAB allocations for SEGMENTED was 12.47 GB vs 12.22 GB (which is a 2% difference and well within sampling error). Looking at the allocations though both had 64% in byte[] alone.

I will run the tests with a LOCAL cache which show more meaningful numbers. The network overhead nullifies the data container itself.

wburns commented:
With a DIST cache with 1 node I saw a 4% degradation, which is probably due to what I said earlier. I will see if there is any way I can trim that off a bit.

```
SEGMENTED

Benchmark          (keyObjectSize)  (nodes)  (useIdentityCache)  (useStrings)  (valueObjectSize)   Mode  Cnt       Score     Error  Units
JMHBenchmarks.put               10        1                true          true               1000  thrpt  100  148785.318 ± 754.433  ops/s

MASTER
Benchmark          (keyObjectSize)  (nodes)  (useIdentityCache)  (useStrings)  (valueObjectSize)   Mode  Cnt       Score      Error  Units
JMHBenchmarks.put               10        1                true          true               1000  thrpt  100  154839.369 ± 1114.430  ops/s
```

wburns commented:
Unfortunately all of my previous runs have been invalidated as I didn't realize I had a store configured, oops!

A new run with single node DIST showed SEGMENTED actually faster in the run, so my guess is that they should be pretty much the same. I have to verify allocation rates still and radargun perf test should be done soon.

```
SEGMENT

Benchmark          (keyObjectSize)  (nodes)  (useIdentityCache)  (useStrings)  (valueObjectSize)   Mode  Cnt        Score      Error  Units
JMHBenchmarks.put               10        1                true          true               1000  thrpt   20  1170006.851 ± 9396.602  ops/s


MASTER

Benchmark          (keyObjectSize)  (nodes)  (useIdentityCache)  (useStrings)  (valueObjectSize)   Mode  Cnt        Score       Error  Units
JMHBenchmarks.put               10        1                true          true               1000  thrpt   20  1168134.261 ± 37600.542  ops/s
```

wburns commented:
Performance job shows no major regressions, however I will see if I can tweak a few things.

wburns commented:
Running with gc profiling with jmh shows better results actually, which is interesting

```
SEGMENT

Benchmark                                           (keyObjectSize)  (nodes)  (useIdentityCache)  (useStrings)  (valueObjectSize)   Mode  Cnt        Score      Error   Units
JMHBenchmarks.put                                                10        1                true          true               1000  thrpt   20  1198320.749 ± 3089.854   ops/s
JMHBenchmarks.put:·gc.alloc.rate                                 10        1                true          true               1000  thrpt   20      682.244 ±    2.062  MB/sec
JMHBenchmarks.put:·gc.alloc.rate.norm                            10        1                true          true               1000  thrpt   20      896.038 ±    0.114    B/op
JMHBenchmarks.put:·gc.churn.PS_Eden_Space                        10        1                true          true               1000  thrpt   20      673.153 ±  346.253  MB/sec
JMHBenchmarks.put:·gc.churn.PS_Eden_Space.norm                   10        1                true          true               1000  thrpt   20      884.454 ±  454.944    B/op
JMHBenchmarks.put:·gc.churn.PS_Survivor_Space                    10        1                true          true               1000  thrpt   20        0.008 ±    0.014  MB/sec
JMHBenchmarks.put:·gc.churn.PS_Survivor_Space.norm               10        1                true          true               1000  thrpt   20        0.011 ±    0.018    B/op
JMHBenchmarks.put:·gc.count                                      10        1                true          true               1000  thrpt   20       15.000             counts
JMHBenchmarks.put:·gc.time                                       10        1                true          true               1000  thrpt   20       19.000                 ms

MASTER

Benchmark                                           (keyObjectSize)  (nodes)  (useIdentityCache)  (useStrings)  (valueObjectSize)   Mode  Cnt        Score      Error   Units
JMHBenchmarks.put                                                10        1                true          true               1000  thrpt   20  1178672.526 ± 7117.907   ops/s
JMHBenchmarks.put:·gc.alloc.rate                                 10        1                true          true               1000  thrpt   20      689.053 ±    4.155  MB/sec
JMHBenchmarks.put:·gc.alloc.rate.norm                            10        1                true          true               1000  thrpt   20      920.038 ±    0.116    B/op
JMHBenchmarks.put:·gc.churn.PS_Eden_Space                        10        1                true          true               1000  thrpt   20      717.780 ±  319.745  MB/sec
JMHBenchmarks.put:·gc.churn.PS_Eden_Space.norm                   10        1                true          true               1000  thrpt   20      959.782 ±  427.584    B/op
JMHBenchmarks.put:·gc.churn.PS_Survivor_Space                    10        1                true          true               1000  thrpt   20        0.007 ±    0.013  MB/sec
JMHBenchmarks.put:·gc.churn.PS_Survivor_Space.norm               10        1                true          true               1000  thrpt   20        0.010 ±    0.018    B/op
JMHBenchmarks.put:·gc.count                                      10        1                true          true               1000  thrpt   20       16.000             counts
JMHBenchmarks.put:·gc.time                                       10        1                true          true               1000  thrpt   20       19.000                 ms
```

wburns commented:
I have tweaked my tests to run DIST in a single node, which is the worst case for segment container. I found that calculating the segment is an operation that can be cached.

By caching the segment at the command level we can prevent additional per entry allocations. It does increase single key write operations wire size by 1-2 bytes. But by caching I found significant improvements making it faster than master for puts.

```
SEGMENT

Benchmark          (keyObjectSize)  (nodes)  (useIdentityCache)  (useStrings)  (valueObjectSize)   Mode  Cnt        Score      Error  Units
JMHBenchmarks.put               10        1                true          true               1000  thrpt   20  1020357.405 ± 5376.089  ops/s
JMHBenchmarks.put               10        1                true          true               1000  thrpt   20  1021888.789 ± 3508.001  ops/s
JMHBenchmarks.put               10        1                true          true               1000  thrpt   20  1010196.212 ± 16508.391  ops/s

SEGMENT no hash

Benchmark          (keyObjectSize)  (nodes)  (useIdentityCache)  (useStrings)  (valueObjectSize)   Mode  Cnt        Score      Error  Units
JMHBenchmarks.put               10        1                true          true               1000  thrpt   20  1294561.360 ± 4250.884  ops/s
JMHBenchmarks.put               10        1                true          true               1000  thrpt   20  1363422.879 ± 6113.529  ops/s
JMHBenchmarks.put               10        1                true          true               1000  thrpt   20  1368859.427 ± 9537.461  ops/s


MASTER

Benchmark          (keyObjectSize)  (nodes)  (useIdentityCache)  (useStrings)  (valueObjectSize)   Mode  Cnt        Score      Error  Units
JMHBenchmarks.put               10        1                true          true               1000  thrpt   20  1184932.260 ± 9514.626  ops/s
JMHBenchmarks.put               10        1                true          true               1000  thrpt   20  1158759.297 ± 7233.317  ops/s
JMHBenchmarks.put               10        1                true          true               1000  thrpt   20  1161191.874 ± 6951.269  ops/s
```

wburns commented:
run performance tests please library

wburns commented:
I also was looking at some JMC reports and found that locking was causing a large amount of allocations. By enabling lock striping and setting lock stripe to 1024, I found another sizable increase in DIST single node performance.

```
SEGMENT no hash striping

Benchmark          (keyObjectSize)  (nodes)  (useIdentityCache)  (useStrings)  (valueObjectSize)   Mode  Cnt        Score       Error  Units
JMHBenchmarks.put               10        1                true          true               1000  thrpt   20  1499331.206 ± 29098.375  ops/s
JMHBenchmarks.put               10        1                true          true               1000  thrpt   20  1477428.516 ± 53062.389  ops/s
JMHBenchmarks.put               10        1                true          true               1000  thrpt   20  1496069.238 ± 33560.358  ops/s
```

With both segment caching and lock striping this is almost a 30% increase in throughput for DIST single node writes. Granted this number will drop dramatically when running in a cluster.

And also despite running more operations I saw a 33% drop in TLAB allocations per second in the JFR output. The TLAB reduction was mostly in part due to the lock striping.

wburns commented:
I found some more allocations especially around notifications that aren't used to trim that saved a a little more in the DIST 1 node case.

```
JMHBenchmarks.put               10        1                true          true               1000  thrpt   20  1555057.024 ± 8951.852  ops/s
JMHBenchmarks.put               10        1                true          true               1000  thrpt   20  1561796.135 ± 23440.237  ops/s
JMHBenchmarks.put               10        1                true          true               1000  thrpt   20  1568670.377 ± 21435.578  ops/s
```

wburns commented:
run performance tests please library

wburns commented:
run performance tests please
library
[UDP 1kB]

wburns commented:
Performance looks really good for non tx micro benchmarks and good on radargun. However it looks like tx might need some additional help (unfortunately the segment caching is done on single key commands which is lost during the commit). I will see if there is any way I can help speed up tx.

diegolovison commented:
Hi William, I saw that you asked for `[UDP 1kB]`
The last time that we ran this configuration was in Mar 20th.
I will run our regression tests against `[UDP 1kB]`. Once it finished, we can see the real values in report. In this case, you don't need to ask to rerun again the performance for this PR, the data will be automatically updated into the report.

We have an internal discussion to improve this.

diegolovison commented:
run performance tests please
library
[UDP 1kB]

wburns commented:
With my micro benchmarks non tx and tx both should be faster for puts now. But we can see what radargun report says.

I also took some time to test out the iteration performance using the benchmark at https://github.com/infinispan/infinispan-benchmarks

```
MASTER
Benchmark                                 (batchSize)  (entryAmount)  (keyObjectSize)  (nodes)  (useIdentityCache)  (useStrings)  (valueObjectSize)   Mode  Cnt   Score   Error  Units
JMHBenchmarks.iteratorParallel                   4096          50000               10        6                true         false               1000  thrpt   20  14.337 ± 0.406  ops/s
JMHBenchmarks.iteratorSequential                 4096          50000               10        6                true         false               1000  thrpt   20   7.832 ± 0.159  ops/s
JMHBenchmarks.noRehashIteratorParallel           4096          50000               10        6                true         false               1000  thrpt   20  15.471 ± 0.599  ops/s
JMHBenchmarks.noRehashIteratorSequential         4096          50000               10        6                true         false               1000  thrpt   20   8.487 ± 0.121  ops/s

JMHBenchmarks.performSize         4096          50000               10        6                true         false               1000  thrpt   20  231.989 ± 5.150  ops/s

SEGMENTED
JMHBenchmarks.iteratorParallel                   4096          50000               10        6                true         false               1000  thrpt   20  14.998 ± 0.339  ops/s
JMHBenchmarks.iteratorSequential                 4096          50000               10        6                true         false               1000  thrpt   20   8.777 ± 0.169  ops/s
JMHBenchmarks.noRehashIteratorParallel           4096          50000               10        6                true         false               1000  thrpt   20  16.212 ± 0.409  ops/s
JMHBenchmarks.noRehashIteratorSequential         4096          50000               10        6                true         false               1000  thrpt   20   9.551 ± 0.191  ops/s


JMHBenchmarks.performSize         4096          50000               10        6                true         false               1000  thrpt   20  682.293 ± 23.666  ops/s
```

The iteration parts showed a little improvement (which was expected), but that is mostly due to network overhead. Whereas size method saw 194% improvement in throughput. I was hoping for 90% at most, so this is quite awesome. I will run parallel stream size as well to see if that differs.

wburns commented:
run performance tests please
library
[UDP 1kB]

wburns commented:
run performance tests please
library

wburns commented:
run performance tests please
library
[UDP 1kB]

diegolovison commented:
run performance tests please
library
[UDP 1kB]

wburns commented:
run performance tests please
library

wburns commented:
run performance tests please
library repl reads

wburns commented:
run performance tests please
library repl reads

wburns commented:
run performance tests please
library repl reads

wburns commented:
Side note but if stats are enabled it hampers REPL read throughput severely, about 45% decrease in throughput. Making them unavailable helps a little above disabling, but not as much of a difference.

wburns commented:
run performance tests please

wburns commented:
The resulting performance numbers while faster are still disappointing compared to what I saw with JMH. 

This is the performance I am seeing with JMH with 100% read hits in REPL

![chart](https://user-images.githubusercontent.com/648175/40576196-7a8d1d26-60bf-11e8-81f2-38bf78462f33.png)

Keep in mind my box is a quad core, hyper threaded.

danberindei commented:
PerfAck machines are Opterons, that might be the reason for the differences you're seeing -- I remember seeing a big improvements from `StripedCounters` on those machines and basically no difference on my machine. I suggest running JMH on those machines and trying to replicate the PerfAck results.

wburns commented:
I ran this on a single node (edg-perf-01) and got similar results with JMH

![chart](https://user-images.githubusercontent.com/648175/40669725-9b060112-6335-11e8-972c-36c50e7aae88.png)

wburns commented:
run performance tests please

ryanemerson commented on core/src/main/java/org/infinispan/commands/write/ApplyDeltaCommand.java:
```diff
@@ -25,7 +25,7 @@ public ApplyDeltaCommand() {
 
    public ApplyDeltaCommand(Object deltaAwareValueKey, Delta delta, Collection<Object> keys, CommandInvocationId commandInvocationId) {
-      super(deltaAwareValueKey, 0, commandInvocationId);
+      super(deltaAwareValueKey, -1, 0, commandInvocationId);
```
Will `segment = -1` not always throw an `IllegalArgumentException` in `AbstractDataCommand`?

wburns replied:
yes, tbh this code is all dead anyways as we don't use it anymore. I wasn't that worried about it.

wburns replied:
`AdvancedCache.applyDelta` doesn't even use the `ApplyDeltaCommand` :D 

https://github.com/infinispan/infinispan/blob/master/core/src/main/java/org/infinispan/cache/impl/CacheImpl.java#L1018

wburns replied:
And the constructor is never invoked and is in a private package. I can add a segment argument if you guys wanted, since it doesn't really matter.

danberindei replied:
Ah yes, I guess it's still around only because it's in the signature of `Visitor.visitApplyDeltaCommand()` and the custom interceptors. Since all the methods already throw `UnsupportedOperationException`, I think we should also throw USO in the constructor, and explain in the javadoc why the class is still around, because to me "deprecated" means it's not recommended, but it still works if you need it.

ryanemerson commented on core/src/main/java/org/infinispan/container/BoundedSegmentedDataContainer.java:
```diff
@@ -0,0 +1,255 @@
+      entries = evictionCache.asMap();
+   }
+
+   public BoundedSegmentedDataContainer(int numSegments, long thresholdSize,
```
Constructor never used

wburns replied:
This is more for @pferraro if needs it :D

ryanemerson commented on core/src/main/java/org/infinispan/container/BoundedSegmentedDataContainer.java:
```diff
@@ -0,0 +1,255 @@
+   @Override
+   public void clear() {
+      entries.clear();
+      // TODO: find out if above clear removes these as well
```
Is this to be resolved in this PR? Or future work?

wburns replied:
I can try real quick.

wburns replied:
Removed it. Also looked at caffeine code and they do invoke the `CacheWriter`.

ryanemerson commented on core/src/main/java/org/infinispan/container/BoundedSegmentedDataContainer.java:
```diff
@@ -0,0 +1,255 @@
+
+   @Override
+   public void removeSegments(IntSet segments) {
+      // Call super remove segments so the maps are remove more efficiently
```
Typo `s/remove/removed`

ryanemerson commented on core/src/main/java/org/infinispan/container/BoundedSegmentedDataContainer.java:
```diff
@@ -0,0 +1,255 @@
+   @Override
+   public long evictionSize() {
+      Policy.Eviction<K, InternalCacheEntry<K, V>> evict = eviction();
+      return evict.weightedSize().orElse(entries.size());
```
Is there are a need for the `evict` variable on the above three methods?

wburns replied:
There isn't specifically. This is how it was before. And I believe I did it that way before, because the `eviction` method can very easily throw a `UnsupportedOperationException` and I wanted to keep the line numbers separate.

We can merge it together though to save the 3 lines if we wanted.

ryanemerson commented on core/src/main/java/org/infinispan/container/DefaultSegmentedDataContainer.java:
```diff
@@ -0,0 +1,301 @@
+      // This can't be injected due to circular dependency with DistributionManager
+      keyPartitioner = cache.getAdvancedCache().getComponentRegistry().getComponent(KeyPartitioner.class);
+
+      CacheMode mode = cache.getCacheConfiguration().clustering().cacheMode();
```
unused variable

wburns replied:
Removed

ryanemerson commented on core/src/main/java/org/infinispan/container/EvictionSetup.java:
```diff
@@ -0,0 +1,53 @@
+ * @author wburns
+ * @since 9.3
+ */
+public class EvictionSetup {
```
I might be missing something, but according to my IDE this class and it's methods are never used.

wburns replied:
Yeah this was refactored away when I added the Abstract class. Let me remove this.

ryanemerson commented on core/src/main/java/org/infinispan/container/SegmentedDataContainer.java:
```diff
@@ -0,0 +1,63 @@
+
+   void addRemovalListener(Consumer<Iterable<InternalCacheEntry<K, V>>> listener);
+
+   void removeRemovalListener(Object listener);
```
This never appears to be called.

wburns replied:
Let me add in the reference real quick.

ryanemerson commented on core/src/main/java/org/infinispan/container/entries/ReadCommittedEntry.java:
```diff
@@ -116,6 +118,17 @@ public boolean isNull() {
 
    @Override
    public final void commit(DataContainer container) {
+      realCommit(-1, container);
```
Instead of manually typing `-1`, should we be using `SegmentSpecificCommand.UNKOWN_SEGMENT`?

ryanemerson replied:
This applies to other parts of the code as well

wburns replied:
Whoops thought I got all of them. Let me double check again.

wburns commented on commons/src/main/java/org/infinispan/commons/util/ConcatIterator.java:
```diff
@@ -0,0 +1,66 @@
+   }
+
+   private static void closeIterator(Iterator<?> iter) {
+      if (iter != null && iter instanceof CloseableIterator) {
```
👍 tbh not sure why IntelliJ didn't recommend that to me.

wburns commented on core/src/main/java/org/infinispan/commands/SegmentSpecificCommand.java:
```diff
@@ -0,0 +1,44 @@
+   /**
+    * The value used to symbolize that a segment is unknown. This value cannot be returned from {@link #getSegment()}.
+    */
+   int UNKOWN_SEGMENT = -1;
```
Oops

wburns commented:
Addressed comments.

tristantarrant commented:
Needs a rebase

danberindei commented on core/src/main/java/org/infinispan/container/SegmentedDataContainer.java:
```diff
@@ -0,0 +1,168 @@
+ * @author wburns
+ * @since 9.3
+ */
+public interface SegmentedDataContainer<K, V> extends DataContainer<K, V> {
```
The new interface added a lot of annoying instanceof checks, how about we add all these methods to `DataContainer` + a method `isSegmented()`?

I was actually thinking of making `SegmentedDataContainer` a super-interface of `DataContainer` with only the methods that we really want to keep post 10.0, and adding default implementations that ignore the segment in `DataContainer`, but I think the only migration path for `AdvancedCache.getDataContainer()` would be to drop it...

wburns replied:
This needs to be thought out further. Added it to the list of things.

danberindei commented on core/src/main/java/org/infinispan/statetransfer/CommitManager.java:
```diff
@@ -123,17 +125,22 @@ public final void commit(final CacheEntry entry, final Flag operation,
+         if (segment != SegmentSpecificCommand.UNKNOWN_SEGMENT && entry instanceof ReadCommittedEntry) {
+            ((ReadCommittedEntry) entry).commit(segment, dataContainer);
+         } else {
+            entry.commit(dataContainer);
```
We really need a `ContextEntry` interface where we can add these methods without worrying about exposing them via `AdvancedCache.getCacheEntry()`

wburns replied:
I agree, it is super messy :(

danberindei commented on core/src/main/java/org/infinispan/statetransfer/StateConsumerImpl.java:
```diff
@@ -967,29 +964,31 @@ protected void removeStaleData(final Set<Integer> removedSegments) throws Interr
+         SegmentedDataContainer sdc = (SegmentedDataContainer) dataContainer;
+         sdc.removeSegments(SmallIntSet.from(removedSegments));
+      } else {
+         dataContainer.executeTask(KeyFilter.ACCEPT_ALL_FILTER, (o, ice) -> {
```
`forEach`?

Off-topic, please make `removedSegments` a `SmallIntSet`.

wburns replied:
Done

danberindei commented on commons/src/main/java/org/infinispan/commons/util/SmallIntSet.java:
```diff
@@ -87,24 +87,17 @@ public SmallIntSet(int initialRange) {
+      } else if (set instanceof IntSet) {
          this.bitSet = new BitSet();
-         set.forEach((IntConsumer) bitSet::set);
+         ((IntSet) set).iterator().forEachRemaining((IntConsumer) bitSet::set);
```
`forEach`

wburns replied:
This is what happens when I add this change in parallel :)

danberindei commented on core/src/main/java/org/infinispan/stream/impl/DistributedCacheStream.java:
```diff
@@ -789,6 +789,12 @@ public void forEachOrdered(Consumer<? super R> action) {
+   @Override
+   public CacheStream<R> filterKeySegments(IntSet segments) {
+      segmentsToFilter = segments;
+      return this;
```
Can't we do this in `AbstractCacheStream`?

wburns replied:
Unfortunately, no. The generics are kinda wonky on this due to `CacheStream` extending `Stream` and this guy implementing `BaseStream`.

danberindei commented on core/src/test/java/org/infinispan/container/DefaultSegmentedDataContainerTest.java:
```diff
@@ -0,0 +1,57 @@
+      for (Cache<Object, Object> cache : caches(CACHE_NAME)) {
+         DataContainer dc = TestingUtil.extractComponent(cache, DataContainer.class);
+         if (!(dc instanceof DefaultSegmentedDataContainer)) {
+            throw new AssertionError("Test requires DefaultSegmentedDataContainer impl");
```
`assertTrue`?

wburns replied:
👍

danberindei commented on core/src/test/java/org/infinispan/distribution/rehash/SharedStoreInvalidationDuringRehashTest.java:
```diff
@@ -48,10 +43,6 @@
 
-   private final ConcurrentMap<Integer, ConcurrentMap<Object, AtomicInteger>> invalidationCounts = CollectionFactory.makeConcurrentMap();
-   private final ConcurrentMap<Integer, ConcurrentMap<Object, AtomicInteger>> l1InvalidationCounts = CollectionFactory.makeConcurrentMap();
-   private Map<Object, Integer> previousOwners = Collections.emptyMap();
```
Does this still test anything? :)

wburns replied:
You would know better than me :D Should I remove? Or is there a recommendation as to how I could keep it?

danberindei commented on core/src/test/java/org/infinispan/expiration/impl/ExpirationFunctionalTest.java:
```diff
@@ -125,6 +125,10 @@ public void testExpirationLifespanInExec() throws Exception {
    public void testExpirationMaxIdleInExec() throws Exception {
+      // Clustered caches can return max idle entries to executeTask
+      if (cache.getCacheConfiguration().clustering().cacheMode().isClustered()) {
+         return;
```
IMO if we know `executeTask()` doesn't trigger maxIdle expiration, we should assert that

wburns replied:
Sure.

danberindei commented on core/src/test/java/org/infinispan/persistence/ClusterCacheLoaderTest.java:
```diff
@@ -23,7 +23,7 @@
  */
 @Test(groups = "functional", testName = "persistence.ClusterCacheLoaderTest")
-@InCacheMode({ CacheMode.INVALIDATION_SYNC, CacheMode.DIST_SYNC, CacheMode.REPL_SYNC, CacheMode.SCATTERED_SYNC })
+@InCacheMode({ CacheMode.INVALIDATION_SYNC/*, CacheMode.DIST_SYNC, CacheMode.REPL_SYNC, CacheMode.SCATTERED_SYNC*/ })
```
Unstable? Please add an explanation

wburns replied:
Didn't mean to comment these out, thanks.

danberindei commented on core/src/test/java/org/infinispan/remoting/transport/jgroups/JGroupsTransportTest.java:
```diff
@@ -50,7 +50,7 @@ public void testSynchronousIgnoreLeaversInvocationToNonMembers() throws Exceptio
       JGroupsTransport transport = (JGroupsTransport) manager(0).getTransport();
       long initialMessages = transport.getChannel().getSentMessages();
-      ReplicableCommand command = new ClusteredGetCommand("key", CACHE_NAME, 0);
+      ReplicableCommand command = new ClusteredGetCommand("key", CACHE_NAME, TestingUtil.getSegmentForKey("key", cache(0)), 0);
```
The segment shouldn't matter, we're sending the command to a different cache anyway.

danberindei commented on core/src/test/java/org/infinispan/remoting/transport/jgroups/JGroupsTransportTest.java:
```diff
@@ -64,7 +64,8 @@ public void testInvokeCommandStaggeredToNonMember() throws Exception {
       JGroupsTransport transport = (JGroupsTransport) manager(0).getTransport();
-      ReplicableCommand command = new ClusteredGetCommand("key", ByteString.fromString("cache"), 0);
+      ReplicableCommand command = new ClusteredGetCommand("key", ByteString.fromString("cache"),
+            TestingUtil.getSegmentForKey("key", cache(0)), 0);
```
Off-topic: please change `ByteString.fromString("cache")` to `CACHE_NAME`

danberindei commented on core/src/test/java/org/infinispan/statetransfer/StateTransferPessimisticTest.java:
```diff
@@ -106,4 +110,20 @@ public int checkKey(Object key, Cache... caches) {
       return c;
    }
+
+   protected void waitUntilDataContainerWillBeIteratedOn(final Cache<?, ?> cache, final CyclicBarrier barrier) {
```
Looks more like a `blockDataContainerIteration()` to me ;)

Also, doesn't it block the cache clear during teardown?

danberindei commented on core/src/test/java/org/infinispan/test/TestingUtil.java:
```diff
@@ -338,6 +340,14 @@ public static Object extractField(Class type, Object target, String fieldName) {
+      if (keyPartitioner != null) {
+         return keyPartitioner.getSegment(key);
+      }
+      return SegmentSpecificCommand.UNKNOWN_SEGMENT;
```
I'd rather throw an exception

wburns replied:
Okay

danberindei commented on core/src/test/java/org/infinispan/stream/BaseSetupStreamIteratorTest.java:
```diff
@@ -48,9 +49,12 @@ protected void enhanceConfiguration(ConfigurationBuilder builder) {
-            cacheMode.isScattered() ? new TestScatteredConsistentHashFactory() : new TestDefaultConsistentHashFactory();
-      builderUsed.clustering().cacheMode(cacheMode).hash().numSegments(3).consistentHashFactory(chf);
+      HashConfigurationBuilder hashConfiguration = builderUsed.clustering().cacheMode(cacheMode).hash().numSegments(3);
+      if (!cacheMode.isReplicated()) {
```
I don't like the idea of having a random CH in the repl test, but if it works I guess it's fine.

wburns replied:
The problem is that the test assigns owners where not all are owned in a REPL cache. So it ends  up being the same issue as wait for initial state transfer.

danberindei replied:
Sure, I didn't mean we should use the existing `TestDefaultConsistentHashFactory`. Looking back at [ISPN-7997](https://issues.jboss.org/browse/ISPN-7997) the default `SyncReplConsistentHashFactory` will allocate one node to be primary owner of each segment so it should be good enough, I would have just preferred a more deterministic implementation similar to `TestDefaultConsistentHashFactory`.

wburns commented on core/src/main/java/org/infinispan/distribution/LocalizedCacheTopology.java:
```diff
@@ -130,6 +138,13 @@ public boolean isReadOwner(Object key) {
+      if (allLocal) {
+         return true;
+      }
+      return distributionInfos[segment].isReadOwner();
```
Yep, sorry was copying the others - I can change it over.

wburns commented on core/src/main/java/org/infinispan/factories/InternalCacheFactory.java:
```diff
@@ -167,4 +209,296 @@ private void bootstrap(String cacheName, AdvancedCache<?, ?> cache, Configuratio
+      @Inject private CacheNotifier<K, V> cacheNotifier;
+
+      // The hasListeners is commented out until EncoderCache can properly pass down the addListener invocation
+      // to the next delegate in the chain. Otherwise we miss listener additions and don't fire events properly.
```
Yeah when I tested it was a slight increase, nothing major. This is actually one of the reasons I changed the visit notify to allow it to be inlined easier.

wburns replied:
I created https://issues.jboss.org/browse/ISPN-9240. I will add to comments in code.

wburns commented on core/src/main/java/org/infinispan/factories/InternalCacheFactory.java:
```diff
@@ -167,4 +209,296 @@ private void bootstrap(String cacheName, AdvancedCache<?, ?> cache, Configuratio
+      }
+   }
+
+   private static class GetReplCache<K, V> extends AbstractGetAdvancedCache<K, V, GetReplCache<K, V>> {
```
It should yes, I mention on the original https://issues.jboss.org/browse/ISPN-9212 that we could even do these for DIST/scattered in a different way. I would say we can add that in later if need though.

danberindei replied:
You can probably do it for dist/scattered as well if you check the segment first - otherwise you might hit a stale entry that you no longer own but haven't removed yet after the last topology update. But we'd probably have to do it within `CacheImpl`, where you can compute the segment, instead of a `CacheImpl` wrapper.

I called out local because the conditions and implementation should be exactly the same.

wburns replied:
Yeah I agree it should be completely the same for LOCAL/INVAL.

wburns commented on core/src/main/java/org/infinispan/commands/CommandsFactoryImpl.java:
```diff
@@ -205,27 +206,37 @@
+         return keyPartitioner.getSegment(key);
+      }
+      // LOCAL and INVAL don't require segments, so just use 0
+      return 0;
```
Sure let me try

wburns commented on core/src/main/java/org/infinispan/commands/CommandsFactory.java:
```diff
@@ -353,15 +354,16 @@
     * @return a ClusteredGetCommand
     */
-   ClusteredGetCommand buildClusteredGetCommand(Object key, long flagsBitSet);
+   ClusteredGetCommand buildClusteredGetCommand(Object key, int segment, long flagsBitSet);
```
Sure, I can try to do this afterwards in another JIRA.

wburns replied:
I created https://issues.jboss.org/browse/ISPN-9242 to handle some of the rework comments I can't get to right away.

wburns commented on core/src/main/java/org/infinispan/commands/SegmentSpecificCommand.java:
```diff
@@ -0,0 +1,44 @@
+      if (command instanceof SegmentSpecificCommand) {
+         return ((SegmentSpecificCommand) command).getSegment();
+      }
+      return UNKNOWN_SEGMENT;
```
I agree re: `UNKNOWN_SEGMENT`

wburns commented on core/src/main/java/org/infinispan/commands/read/AbstractDataCommand.java:
```diff
@@ -6,26 +6,43 @@
    private long flags;
+   // These 2 ints have to stay next to each other to ensure they are aligned together
    private int topologyId = -1;
+   private int segment = -1;
```
sure

wburns commented on core/src/main/java/org/infinispan/commands/read/EntrySetCommand.java:
```diff
@@ -156,16 +157,26 @@ public boolean remove(Object o) {
       public CacheStream<CacheEntry<K, V>> stream() {
-         return new LocalCacheStream<>(new EntryStreamSupplier<>(cache, getSegmentMapper(cache),
-                 super::stream), false, cache.getAdvancedCache().getComponentRegistry());
+         return stream(false);
```
Okay

wburns commented on core/src/main/java/org/infinispan/container/BoundedSegmentedDataContainer.java:
```diff
@@ -0,0 +1,253 @@
+      int segment = getSegmentForKey(key);
+      ConcurrentMap<K, InternalCacheEntry<K, V>> map = BoundedSegmentedDataContainer.super.getMapForSegment(segment);
+      if (map != null) {
+         map.put(key, value);
```
Well the concern is both the segment id computation and the fact that we have to iterate over all the non owned entries. The combination of both is why for the size method I saw almost 3x increase in throughput.

danberindei replied:
"iterate overa all the non owned entries" -- why would there be any non owned entries in the data container?

wburns replied:
Sorry I meant non primary owned entries. And there would be non owned entries with L1. Also this allows faster iteration on retries when we finish a subset of segments.

wburns commented on core/src/main/java/org/infinispan/container/DataContainer.java:
```diff
@@ -90,10 +90,19 @@
 
    /**
-    *
+    * Returns how many entries are present in the data container without counting entries that are currently expired
```
Hrmm I can do that as a separate commit, sure.

wburns commented on core/src/main/java/org/infinispan/container/DataContainer.java:
```diff
@@ -90,10 +90,19 @@
+   default int size() {
+      int size = 0;
+      // We have to loop through to make sure to remove expired entries
+      for (InternalCacheEntry<K, V> ignore : this) {
```
The problem I found with that is I have to use an `AtomicInteger` instead of int and it can't early terminate due to having more than `Integer.MAX_VALUE` entries.

wburns commented on core/src/main/java/org/infinispan/container/DataContainer.java:
```diff
@@ -175,21 +196,51 @@
     */
-   void executeTask(final KeyFilter<? super K> filter, BiConsumer<? super K, InternalCacheEntry<K, V>> action) throws InterruptedException;
+   default void executeTask(final KeyFilter<? super K> filter, BiConsumer<? super K, InternalCacheEntry<K, V>> action) throws InterruptedException {
+      iterator().forEachRemaining(ice -> {
```
+1

wburns commented on core/src/main/java/org/infinispan/container/AbstractSegmentedDataContainer.java:
```diff
@@ -0,0 +1,634 @@
+         if (trace())
+            log().tracef("Store %s in container", copy);
+
+         entries.compute(copy.getKey(), (key, entry) -> {
```
Hrmm probably, this was just copied from the `DefaultDataContainer`. I will try to ensure we aren't calling the trace method inside the `compute` method. This can be done afterwards though, as I don't want to toy with this right now.

wburns commented on core/src/main/java/org/infinispan/container/AbstractSegmentedDataContainer.java:
```diff
@@ -0,0 +1,634 @@
+      Iterator<InternalCacheEntry<K, V>> iter = iteratorIncludingExpired(segments);
+      while (iter.hasNext()) {
+         iter.next();
+         iter.remove();
```
Hrmm that should work. Guessing I just didn't think about it too hard :)

wburns replied:
Ah it was because of `DefaultDataContainer` - I will change this and try to have `DefaultDataContainer` no longer implement `SegmentedDataContainer` afterwards as mentioned above.

wburns commented on core/src/main/java/org/infinispan/container/AbstractSegmentedDataContainer.java:
```diff
@@ -0,0 +1,634 @@
+   @Inject protected Cache<K, V> cache;
+   protected KeyPartitioner keyPartitioner;
+
+   protected final List<Consumer<Iterable<InternalCacheEntry<K, V>>>> listeners = new CopyOnWriteArrayList<>();
```
At this point, no. Seemed less of a headache to just implement it fully.

wburns commented on core/src/main/java/org/infinispan/container/AbstractSegmentedDataContainer.java:
```diff
@@ -0,0 +1,634 @@
+                  initializedTime = true;
+               }
+               if (!entry.isExpired(now) || expirationManager.entryExpiredInMemoryFromIteration(entry, now).join()
+                     == Boolean.FALSE) {
```
The idea was if we allowed for some sort of non blocking variant for iteration (which we don't have a non blocking iteration yet). We can discuss this in a different JIRA maybe?

danberindei replied:
Sure, I just want to make sure we don't implement something "just in case" and when we get to actually using it we get stuck doing it the same way because "we've always done it this way"

wburns commented on core/src/main/java/org/infinispan/container/AbstractSegmentedDataContainer.java:
```diff
@@ -0,0 +1,634 @@
+
+      @Override
+      public void forEachRemaining(Consumer<? super InternalCacheEntry<K, V>> action) {
+         // We don't call the forEachRemaining on the actual spliterator since, we want to keep the time between
```
K

wburns commented on core/src/main/java/org/infinispan/container/AbstractSegmentedDataContainer.java:
```diff
@@ -0,0 +1,634 @@
+            if (currentEntry.canExpire()) {
+               if (!initializedTime) {
+                  now = timeService.wallClockTime();
+                  initializedTime = true;
```
Okay, just another copy/paste casualty :)

wburns commented on core/src/main/java/org/infinispan/container/DefaultDataContainer.java:
```diff
@@ -379,287 +217,27 @@ public long evictionSize() {
-            return false;
+         if (removedEntries != null) {
+            List<InternalCacheEntry<K, V>> unmod = Collections.unmodifiableList(removedEntries);
+            listeners.forEach(c -> c.accept(unmod));
```
Hrmm I would have to think of the ramifications of that, in that case I assume you expect the listener to provide the segment that maps to a given set of entries then as well?

danberindei replied:
No, the way I imagine it is the listener would get only a segment id and the interceptor would reset its count to 0 for that segment.

wburns replied:
Ah, yes that is an interesting approach and actually would work better, especially for off heap.

wburns commented on core/src/main/java/org/infinispan/container/AbstractSegmentedDataContainer.java:
```diff
@@ -0,0 +1,634 @@
+   protected abstract int getSegmentForKey(Object key);
+
+   protected abstract Log log();
+   protected abstract boolean trace();
```
Did you want me to log all the messages in `AbstractSegmentedDataContainer` at the `AbstractSegmentedDataContainer` log?

wburns commented on core/src/main/java/org/infinispan/container/AbstractSegmentedDataContainer.java:
```diff
@@ -0,0 +1,634 @@
+ * @author wburns
+ * @since 9.3
+ */
+public abstract class AbstractSegmentedDataContainer<K, V> implements DataContainer<K, V>, SegmentedDataContainer<K, V> {
```
Yeah I originally had them separate, which it looks like I will try to go back to :)

danberindei replied:
I hope it's not because of our discussion, because that's not what I meant :)

wburns replied:
Gotcha.

wburns commented on core/src/main/java/org/infinispan/container/DefaultDataContainer.java:
```diff
@@ -318,48 +179,25 @@ public void clear() {
-   public Spliterator<InternalCacheEntry<K, V>> spliterator() {
-      return new EntrySpliterator(entries.values().spliterator());
+   public Spliterator<InternalCacheEntry<K, V>> spliteratorIncludingExpired() {
+      // Technically this spliterator is distinct, but it won't be set - we assume that is okay for now
```
The only reason I made sure it was there, was because at some point I will be making the stream stuff care more about what flags are set and DISTINCT is just one more :D

wburns commented on core/src/main/java/org/infinispan/container/EntryFactoryImpl.java:
```diff
@@ -213,8 +216,24 @@ private CacheEntry getFromContainer(Object key, boolean isOwner, boolean writeOp
-   private CacheEntry getFromContainerForRead(Object key, boolean isOwner) {
-      InternalCacheEntry ice = container.get(key);
+   private InternalCacheEntry peek(Object key, int segment) {
+      if (segment != SegmentSpecificCommand.UNKNOWN_SEGMENT && container instanceof SegmentedDataContainer) {
```
This would require all places that invoke `EntryFactory` to provide a 0 or greater number, this can be added later maybe.

wburns commented on core/src/main/java/org/infinispan/container/L1DefaultSegmentedDataContainer.java:
```diff
@@ -0,0 +1,165 @@
+
+/**
+ * Segmented data container that also allows for non owned segments to be written to a temporary map (L1). This
+ * map is cleared whenever a segment becomes no longer owned.
```
Not having a separate implementation you mean always having the temporary map available?

Also I was just taking the easy route and whenever a segment is removed. I can move this to the topology changed event instead maybe.

danberindei replied:
No, I meant using a `DefaultSegmentedDataContainer` even when L1 is enabled.

wburns replied:
So basically making the `L1SegmentedDataContainer` the `DefaultSegmentedContainer`?

wburns commented on core/src/main/java/org/infinispan/container/SegmentedDataContainer.java:
```diff
@@ -0,0 +1,168 @@
+    * Removes entries from the container whose key maps to one of the provided segments
+    * @param segments segments of entries to remove
+    */
+   void clear(IntSet segments);
```
`clear(IntSet)` just removes the entries where as `removeSegments(IntSet)` is the method invoked when you no longer own those segments. They require different behaviors.

wburns commented on core/src/main/java/org/infinispan/container/SegmentedDataContainer.java:
```diff
@@ -0,0 +1,168 @@
+   /**
+    * Removes and un-associates the given segments. This will notify any listeners registered via
+    * {@link #addRemovalListener(Consumer)}. There is no guarantee if the consumer is invoked once or multiple times
+    * for a given group of segments and could be in any order
```
Not yet, no. :)

wburns commented on core/src/main/java/org/infinispan/container/entries/ReadCommittedEntry.java:
```diff
@@ -116,6 +118,17 @@ public boolean isNull() {
 
    @Override
    public final void commit(DataContainer container) {
+      realCommit(SegmentSpecificCommand.UNKNOWN_SEGMENT, container);
```
I will see if I can remove it in the refactoring.

wburns commented on core/src/main/java/org/infinispan/container/DefaultSegmentedDataContainer.java:
```diff
@@ -0,0 +1,298 @@
+         ConsistentHash endCH = topologyChangedEvent.getWriteConsistentHashAtEnd();
+         if (endCH.getMembers().contains(localNode)) {
+            Set<Integer> segments = endCH.getSegmentsForOwner(localNode);
+            startSegments(segments);
```
Hrmm that is a good idea, instead of relying on the listener.

wburns commented on core/src/main/java/org/infinispan/container/EntryFactoryImpl.java:
```diff
@@ -94,7 +95,8 @@ public void wrapEntryForWriting(InvocationContext ctx, Object key, boolean isOwn
       } else {
          // Not in the context yet.
-         CacheEntry cacheEntry = getFromContainer(key, isOwner, true);
+         // TODO: need to handle this segment
```
Looks like I forgot to remove it.

wburns commented on core/src/main/java/org/infinispan/distribution/LocalizedCacheTopology.java:
```diff
@@ -43,10 +44,17 @@
+   }
+
+   public static LocalizedCacheTopology makeSingletonTopology(CacheMode cacheMode, KeyPartitioner keyPartitioner,
+         int numSegments, Address localAddress) {
```
Okay.

wburns commented on core/src/main/java/org/infinispan/distribution/LocalizedCacheTopology.java:
```diff
@@ -130,6 +138,13 @@ public boolean isReadOwner(Object key) {
       return distributionInfos[segmentId].isReadOwner();
    }
 
+   public boolean isReadOwner(int segment) {
```
Sure

wburns commented on core/src/main/java/org/infinispan/factories/DataContainerFactory.java:
```diff
@@ -29,29 +33,47 @@
-               configuration.memory().size(newSize.get()));
-         configuration.memory().attributes().attribute(MemoryConfiguration.SIZE).addListener((newSize, old) ->
+               memoryConfiguration.size(newSize.get()));
+         memoryConfiguration.attributes().attribute(MemoryConfiguration.SIZE).addListener((newSize, old) ->
```
yeah it will resize the container down if you change the size dynamically.

wburns commented on core/src/main/java/org/infinispan/factories/InternalCacheFactory.java:
```diff
@@ -12,29 +19,43 @@
- * An internal factory for constructing Caches.  Used by the {@link DefaultCacheManager}, this is not intended as public
- * API.
+ * An internal factory for constructing Caches.  Used by the {@link org.infinispan.manager.DefaultCacheManager}, this
+ * is not intended as public API.
```
I wasn't trying to modify anything here, organize import got me :)

wburns commented on core/src/main/java/org/infinispan/factories/InternalCacheFactory.java:
```diff
@@ -54,8 +77,8 @@
-                                  GlobalComponentRegistry globalComponentRegistry,
-                                  String cacheName) throws CacheConfigurationException {
+         GlobalComponentRegistry globalComponentRegistry,
+         String cacheName) throws CacheConfigurationException {
```
tbh I don't like what this did at all.

wburns commented on core/src/main/java/org/infinispan/factories/InternalCacheFactory.java:
```diff
@@ -70,10 +93,29 @@
+      BiFunction<DataConversion, DataConversion, AdvancedCache<K, V>> usedBuilder;
+      // We can optimize REPL reads that meet some criteria. This allows us to bypass
+      if (configuration.clustering().cacheMode().isReplicated() && !configuration.persistence().usingStores()
+            && !configuration.transaction().transactionMode().isTransactional() && configuration.clustering().stateTransfer().awaitInitialTransfer()) {
```
I am not necessarily, the problem is until we wait for state transfer we don't own the segments, which mean REPL has to go remote. We would have to wait until the cache is owner before we start doing the optimizations then. It was a quick fix.

wburns commented on core/src/main/java/org/infinispan/factories/InternalCacheFactory.java:
```diff
@@ -167,4 +209,296 @@ private void bootstrap(String cacheName, AdvancedCache<?, ?> cache, Configuratio
+      @Override
+      public V get(Object key) {
+         assertKeyNotNull(key);
+         checkCanRun(cache, componentRegistry);
```
I changed to `cache.getName`. Unfortunately the whole wiring thing is a rat's nest. I will have to revisit it after.

wburns commented on core/src/main/java/org/infinispan/factories/InternalCacheFactory.java:
```diff
@@ -167,4 +209,296 @@ private void bootstrap(String cacheName, AdvancedCache<?, ?> cache, Configuratio
+   private static class PartitionHandlingCache<K, V> extends AbstractGetAdvancedCache<K, V, PartitionHandlingCache<K, V>> {
+      @Inject private PartitionHandlingManager manager;
+
+      private final long bitFlags;
```
The flags from `withFlags`. That is what they were called there :D

wburns commented on core/src/main/java/org/infinispan/factories/InternalCacheFactory.java:
```diff
@@ -70,10 +93,29 @@
+      final BiFunction<DataConversion, DataConversion, AdvancedCache<K, V>> actualBuilder = (kc, kv) -> new CacheImpl<>(cacheName);
+      BiFunction<DataConversion, DataConversion, AdvancedCache<K, V>> usedBuilder;
+      // We can optimize REPL reads that meet some criteria. This allows us to bypass
+      if (configuration.clustering().cacheMode().isReplicated() && !configuration.persistence().usingStores()
```
I would like to, but didn't on the first pass.

wburns commented on core/src/main/java/org/infinispan/interceptors/distribution/L1NonTxInterceptor.java:
```diff
@@ -279,7 +280,7 @@ private Object handleWriteManyCommand(InvocationContext ctx, WriteCommand comman
       //we also need to remove from L1 the keys that are not ours
       Iterator<VisitableCommand> subCommands = keys.stream().filter(
-            k -> !cdl.getCacheTopology().isWriteOwner(k)).map(k -> removeFromL1Command(ctx, k)).iterator();
+            k -> !cdl.getCacheTopology().isWriteOwner(k)).map(k -> removeFromL1Command(ctx, k, -1)).iterator();
```
+1

wburns replied:
This needs a bit of refactoring, will hold off until later.

wburns commented on core/src/main/java/org/infinispan/interceptors/impl/EntryWrappingInterceptor.java:
```diff
@@ -114,16 +115,22 @@
-            addVersionRead((TxInvocationContext) rCtx, cacheEntry, dataCommand.getKey());
-         }
+         // This invokes another method as this is only done with a specific configuration and we want to inline
+         // the notifier below
```
It is pretty simple and from what I saw it was pretty consistently unable to inline this entire method, because it would make the caller too large. This would help inline when not using repeatable read, which seems good to me.

wburns commented on core/src/main/java/org/infinispan/interceptors/impl/PrefetchInterceptor.java:
```diff
@@ -200,12 +201,13 @@ protected Object handleReadCommand(InvocationContext ctx, DataCommand command) t
       }
-      GetCacheEntryCommand getCacheEntryCommand = commandsFactory.buildGetCacheEntryCommand(key, EnumUtil.bitSetOf(Flag.CACHE_MODE_LOCAL));
-      getCacheEntryCommand.setTopologyId(topologyId);
+      int segment = SegmentSpecificCommand.extractSegment(command);
```
👍

wburns commented on core/src/main/java/org/infinispan/interceptors/locking/ClusteringDependentLogic.java:
```diff
@@ -118,7 +120,18 @@ default Address getPrimaryOwner(Object key) {
+    * @param ctx
+    * @param key
+    * @param segment if 0 or greater assumes the underlying container is segmented. otherwise inserts into container
+    *                without using segments
```
I am not familiar with it either, this was added by someone else. I can log a JIRA to see if we should remove it.

wburns replied:
https://issues.jboss.org/browse/ISPN-9244

wburns commented on core/src/main/java/org/infinispan/marshall/exts/ReplicableCommandExternalizer.java:
```diff
@@ -72,6 +74,9 @@ protected void writeCommandParameters(ObjectOutput output, ReplicableCommand com
          output.writeInt(((TopologyAffectedCommand) command).getTopologyId());
       }
+      if (command instanceof SegmentSpecificCommand) {
+         UnsignedNumeric.writeUnsignedInt(output, ((SegmentSpecificCommand) command).getSegment());
```
Removed

wburns commented on core/src/main/java/org/infinispan/persistence/PersistenceUtil.java:
```diff
@@ -119,11 +121,19 @@ public static InternalMetadata internalMetadata(InternalCacheValue icv) {
+            timeService, isLoaded);
+   }
+
+   // TODO: need to add back in old methods
```
Nope, just forgot to remove this.

wburns commented:
Updated and fixed most of the comments.  I will be working on the others in a bit.

I have a JIRA linking to each of the ones in particular to look at https://issues.jboss.org/browse/ISPN-9242

danberindei commented on core/src/main/java/org/infinispan/commands/remote/ClusteredGetCommand.java:
```diff
@@ -45,6 +46,7 @@
    private InternalEntryFactory entryFactory;
    //only used by extended statistics. this boolean is local.
    private boolean isWrite;
+   private int segment = -1;
```
Initializer is no longer needed

wburns replied:
Okay.

danberindei commented on core/src/main/java/org/infinispan/container/impl/BoundedSegmentedDataContainer.java:
```diff
@@ -0,0 +1,253 @@
+ */
+public class BoundedSegmentedDataContainer<K, V> extends DefaultSegmentedDataContainer<K, V> {
+   protected final Cache<K, InternalCacheEntry<K, V>> evictionCache;
+   protected final ConcurrentMap<K, InternalCacheEntry<K, V>> entries;
```
I'm not sure why you have both `evictionCache` and `entries`, and you also extend `DefaultSegmentedDataContainer`

wburns replied:
The `evictionCache` is required soley to get at the eviction information.
The `entries` is the actual Map we write to for caffeine.
None of these deal with segmented information, the segmented data is still all stored in the maps in the `DefaultSegmentedDataContainer` for iteration and peek purposes.

When we write to the caffeine map we propagate the updates to the underlying segmented containers (essentially indexing that data).

danberindei replied:
Ah ok, I thought the `entries` field was new, I didn't realize you had it from the start

danberindei commented on core/src/main/java/org/infinispan/container/impl/SegmentedDataContainer.java:
```diff
@@ -0,0 +1,169 @@
+ * @author wburns
+ * @since 9.3
+ */
+public interface SegmentedDataContainer<K, V> extends DataContainer<K, V> {
```
I would have liked to merge `SegmentedDataContainer` into `DataContainer` in this PR, not later

wburns replied:
So I thought we wanted to separate the interfaces? Or now you want all the methods on `DataContainer`?

danberindei commented on core/src/main/java/org/infinispan/distribution/ch/impl/SingleSegmentKeyPartitioner.java:
```diff
@@ -0,0 +1,26 @@
+   }
+
+   private static class SingletonHelper {
+      private static final SingleSegmentKeyPartitioner INSTANCE = new SingleSegmentKeyPartitioner();
```
Why hide it in another class?

wburns replied:
This way it is only initialized if we need it.

wburns replied:
Obviously low (and will be retained on the heap until class is unloaded), but why even have to create it if we don't have to :D

danberindei replied:
As always, it's a balance act... a separate class will also use some heap and even more metaspace.

wburns replied:
Yeah for the initialized part, true.

danberindei commented on core/src/main/java/org/infinispan/factories/InternalCacheFactory.java:
```diff
@@ -167,4 +209,296 @@ private void bootstrap(String cacheName, AdvancedCache<?, ?> cache, Configuratio
+   }
+
+   private static class PartitionHandlingCache<K, V> extends AbstractGetAdvancedCache<K, V, PartitionHandlingCache<K, V>> {
+      @Inject private PartitionHandlingManager manager;
```
You already have a reference to the cache, so you don't need to inject the manager

danberindei commented on core/src/main/java/org/infinispan/factories/InternalCacheFactory.java:
```diff
@@ -167,4 +209,296 @@ private void bootstrap(String cacheName, AdvancedCache<?, ?> cache, Configuratio
+      }
+
+      private void interceptorStart() {
+         // This has to be done after the cache is wired - otherwise we can get a circular dependency
```
Yes, it needs the field if it's delegating to `Cache`, but nobody forced us to have those managed operations in the `CacheMgmtInterceptor` MBean :) The cache MBean itself could have provided that information, IMO there's no reason why an interceptor should be exposed directly in JMX.

wburns commented on commons/src/main/java/org/infinispan/commons/util/ComposeArraySpliterator.java:
```diff
@@ -0,0 +1,85 @@
+
+   private Spliterator<E> currentSpliterator;
+
+   public ComposeArraySpliterator(IntFunction<Collection<E>> toCollection, int length, int additionalCharacteristics) {
```
Sure

wburns commented:
Updated

danberindei commented:
Integrated, thanks Will!
